### ISSUE TITLE
`lib.hiPrio` gets erased after calling `override` or `overrideAttrs`

### ISSUE THREAD

**Issue by andrewhamon** [NixOS/nixpkgs#323624]
### Describe the bug

`lib.hiPrio` and related functions update `meta` using [simple merging](https://github.com/NixOS/nixpkgs/blob/cd6602819bb0f13b32d6b923a7c46602ee6591f3/lib/meta.nix#L41-L42).

If a package later has an override applied, though, the updated `meta` is discarded. This can be really surprising, particularly to those less well-versed in nix.

### Steps To Reproduce
Try this in a REPL:

```
(pkgs.lib.hiPrio pkgs.cowsay).meta.priority
# => -10

((pkgs.lib.hiPrio pkgs.cowsay).overrideAttrs { }).meta.priority or "priority is not set"
# => "priority is not set"

# also impacts `override`
((pkgs.lib.hiPrio pkgs.cowsay).override { }).meta.priority or "priority is not set"
# => "priority is not set"
```

### Expected behavior
Functions the "mutate" derivations should do so using `overrideAttrs`.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.5.0, macOS 14.5`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.22.0`
 - nixpkgs: `/nix/store/rhg90jpryc286xn9xjy6qjiaap6pjgdc-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by andrewhamon** [comment]
Similar issue from long ago where this is mentioned: https://github.com/NixOS/nixpkgs/issues/7425

**Comment by eclairevoyant** [comment]
Why not move the hiPrio outside the override?

**Comment by andrewhamon** [comment]
Yes, that's an easy workaround to apply, _if you know what the problem is_. But if you don't know already, I don't think the solution is obvious.

Do you see any downsides to adopting a merge function [such as this](https://github.com/tvlfyi/kit/blob/canon/readTree/default.nix#L69-L78)?

**Comment by infinisil** [comment]
Ping @winterqt who just ran into this too :)

**Comment by eclairevoyant** [comment]
> Yes, that's an easy workaround to apply, _if you know what the problem is_.

I don't consider it a workaround, as conceptually it doesn't make as much sense to put the hiPrio before overriding? Because what you want to be hiPrio'd is the overriden package, not the original package.

**Comment by andrewhamon** [comment]
@eclairevoyant in the narrow case of setting priorities, i might agree with your take. But this impacts all meta attrs specified with `addMetaAttrs`.

I still think it is very surprising, in all cases, that invoking an override will reset fields that were not specified in the override.
